experimental/sql/src/projection_pushdown.py

```diff
@@ -0,0 +1,86 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import (RewritePattern, GreedyRewritePatternApplier,
+                                   PatternRewriteWalker, PatternRewriter,
+                                   op_type_rewrite_pattern)
+
+import dialects.rel_alg as RelAlg
+
+
+@dataclass
+class ProjectionPushdownRewriter(RewritePattern):
+
+  def find_cols_in_expr(self, op: RelAlg.Expression) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its children.
+    """
+    if isinstance(op, RelAlg.BinOp):
+      return self.find_cols_in_expr(op.lhs.op) + self.find_cols_in_expr(
+          op.rhs.op)
+    if isinstance(op, RelAlg.Column):
+      return [op.col_name.data]
+    if isinstance(op, RelAlg.Compare):
+      return self.find_cols_in_expr(op.left.op) + self.find_cols_in_expr(
+          op.right.op)
+    return []
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Project, rewriter: PatternRewriter):
+    if not isinstance(op.input.op, RelAlg.Select):
+      return
+
+    select_cols = [
+        elem for sublist in (
+            self.find_cols_in_expr(o) for o in op.input.op.predicates.ops)
+        for elem in sublist
+    ]
```

Comment:
I think I want to split this into several steps with the idea of peephole rewrites in mind.

First, all kinds of operators (but projections) get a projection inserted just before them mapping away all columns that won't be needed later on.

Second, the exact projections, so the expressions that define the projections, are pushed down as far as possible.

Third, projections, that follow are followed by other projections, are fused into single projections.

Lastly, all projections that are just the identity are removed.

Reply:
I am not quite sure that this is the way we really want to have this in the end, and I am in particular uncertain about what to do with joins, but I think that this could be a nice first step in the right direction. And if these things really work the way I want them to I feel that this might give us a lot of what we need.

Reply:
I have the feeling that what you describe should work. It may not be the most efficient way to do things if you don't pay attention to the order in which patterns apply. (Projections only affect downstream operators, so you can do the rewrite in one pass from root to the leaves but patterns might be applied in the opposite direction, so you'd get quadratic runtime.) But it's definitely OK as a first step and probably fixable by controlling the application order...